### PR TITLE
For phones with file-based encryption, use correct Context

### PR DESCRIPTION
**Pull request by Rijul-A**
Due to file-based encryption being introduced in Nougat, preferences are not available to be read into until the system is unlocked for the first time after each reboot. A workaround for this is to use Device Protected Storage (as opposed to Credential Protected Storage), also called as Direct Boot, which enables the preferences to be stored in decrypted format. 
https://github.com/rovo89/XposedBridge/issues/206
https://github.com/GravityBox/GravityBox/commit/d825b7be445c18de9a90eb8f77c1c5eb2b74deb8

Changes include:
- Constructor of the provider changed to use a map of file names and modes (device protected / credential protected) instead of an array of file names.
- Functions using `Context` in the provider changed to use the correct version of context, based on the mode passed previously.
- README.md updated to explain the change.

### CODE REVIEW

**Comment by Rijul-A** [comment]
For a full explanation of the changes required in-app, and in-Xposed, please see this comment https://github.com/apsun/RemotePreferences/issues/4#issuecomment-420959821

**Comment by apsun** [comment]
Interesting, thanks for this. Could we not break backwards compatibility for existing library users? Keep the `(String authority, String[] prefNames)` constructor, and assume that all preferences in that list do not use device protected storage. Also I would prefer a `class RemotePreferenceFile { String prefFile; boolean isDeviceProtected; }` instead of `Map<String, Boolean>`.

```
public RemotePreferenceProvider(..) {
    this(authority, toPrefFiles(prefNames));
}

public RemotePreferenceProvider(String authority, RemotePreferenceFile[] prefFiles) {
   ...
}

private static RemotePreferenceFile[] toPrefFiles(String[] prefNames) {
    for (...) {
        prefFiles[i] = new RemotePreferenceFile(prefNames[i], /* isDeviceProtected */ false);
    }
}
```

**Comment by Rijul-A** [comment]
That makes sense, I'll get back to you with these changes

**Comment by Rijul-A** [comment]
I was looking at it some more and I realized that the library supports different Android versions, so I used the following check. Not sure if that is okay, could you check before merging please?
```
boolean isN = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N;
if (isN && remotePreferenceFile.isDeviceProtected)
    deviceProtectedContext = context.createDeviceProtectedStorageContext();
else
    deviceProtectedContext = context;
```

File: library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java
Comment (apsun) [inline library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java]: I'd prefer you moved this doc to the "main" constructor below and replace this one with something like:

"See {@link (other constructor)}. All preference files are assumed to be located in credential protected (default) rather than device protected storage"

File: library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java
Comment (apsun) [inline library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java]: Just `RemotePreferenceFile file` is fine here, reduces some of the line splitting below

File: library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java
Comment (apsun) [inline library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java:435]: Let's refactor this and the following loops into a new method `private RemotePreferenceFile getFileForPreferences(SharedPreferences prefs)`. Should make the code a bit easier to read.

File: library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java
Comment (apsun) [inline library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java]: Please explicitly add the accessibility modifiers here (I think `public` would be fine?)

File: library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java
Comment (apsun) [inline library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java]: Same (this one can be `private` I believe)

File: README.md
Comment (apsun) [inline README.md]: "Device encrypted preferences" would be a better title here

File: README.md
Comment (apsun) [inline README.md:131]: I don't think this section is relevant to RemotePreferences, if someone is using device encrypted storage they should already have done this elsewhere in their application. But it's your call, I'm not very familiar with the new API.

File: README.md
Comment (apsun) [inline README.md]: Minor nitpick: array line break style should be
```
super("com.example.app.preferences", new RemotePreferenceFile[] {
    new RemotePreferenceFile("main_prefs", true)
});
```

**Review by apsun — CHANGES_REQUESTED** [review]
Sorry for taking so long to get back to this, overall the changes look pretty good, just have some minor stylistic changes. Thanks!

File: library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java
Comment (apsun) [inline library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java]: Please use braces around all conditions, and spaces around operators

```
if (remotePreferenceFile == null) {
    return;
}
````

File: library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java
Comment (apsun) [inline library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java]: Same, add braces here and in the loop below

File: library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java
Comment (apsun) [inline library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java]: Braces

File: library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java
Comment (apsun) [inline library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java]: Try to avoid really complex ternary conditionals. This should be

```
if (isDeviceProtected && Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
    context = context.createDeviceProtectedStorageContext();
}
ContentResolver resolver = context.getContentResolver();
```

File: library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java
Comment (apsun) [inline library/src/main/java/com/crossbowffs/remotepreferences/RemotePreferenceProvider.java]: Fields should be `private final` and the name should be prefixed with `m` (i.e. `mPrefFileName`, `mIsDeviceProtected`) as per Android coding style.

**Comment by Rijul-A** [comment]
Hey @apsun, checking in if all is good with this now?

**Comment by Rijul-A** [comment]
Possibly a solution for #1 too

**Review by apsun — CHANGES_REQUESTED** [review]
Sorry, forgot to submit this last week. Can you just fix these minor nitpicks? Other than that everything looks good!

**Review by apsun — APPROVED** [review]

**Comment by apsun** [comment]
Merged, thank you!

**Pull request merged**
